src/iter/par_bridge.rs

```diff
@@ -0,0 +1,198 @@
+use crossbeam_deque::{Deque, Stealer, Steal};
+
+use std::thread::yield_now;
+use std::sync::{Mutex, TryLockError};
+use std::sync::atomic::{AtomicUsize, AtomicBool, Ordering};
+
+use iter::ParallelIterator;
+use iter::plumbing::{UnindexedConsumer, UnindexedProducer, bridge_unindexed, Folder};
+use current_num_threads;
+
+/// Conversion trait to convert an `Iterator` to a `ParallelIterator`.
+///
+/// This creates a "bridge" from a sequential iterator to a parallel one, by distributing its items
+/// across the Rayon thread pool. This has the advantage of being able to parallelize just about
+/// anything, but the resulting `ParallelIterator` can be less efficient than if you started with
+/// `par_iter` instead. However, it can still be useful for iterators that are difficult to
+/// parallelize by other means, like channels or file or network I/O.
+///
+/// The resulting iterator is not guaranteed to keep the order of the original iterator.
+///
+/// # Examples
+///
+/// To use this trait, take an existing `Iterator` and call `par_bridge` on it. After that, you can
+/// use any of the `ParallelIterator` methods:
+///
+/// ```
+/// use rayon::iter::ParallelBridge;
+/// use rayon::prelude::ParallelIterator;
+/// use std::sync::mpsc::channel;
+///
+/// let rx = {
+///     let (tx, rx) = channel();
+///
+///     tx.send("one!");
+///     tx.send("two!");
+///     tx.send("three!");
+///
+///     rx
+/// };
+///
+/// let mut output: Vec<&'static str> = rx.into_iter().par_bridge().collect();
+/// output.sort_unstable();
+///
+/// assert_eq!(&*output, &["one!", "three!", "two!"]);
+/// ```
+pub trait ParallelBridge {
+    /// What is the type of the output `ParallelIterator`?
+    type Iter: ParallelIterator<Item = Self::Item>;
+
+    /// What is the `Item` of the output `ParallelIterator`?
+    type Item: Send;
+
+    /// Convert this type to a `ParallelIterator`.
+    fn par_bridge(self) -> Self::Iter;
+}
+
+impl<T: Iterator + Send> ParallelBridge for T
+    where T::Item: Send
+{
+    type Iter = IterParallel<T>;
+    type Item = T::Item;
+
+    fn par_bridge(self) -> Self::Iter {
+        IterParallel {
+            iter: self,
+        }
+    }
+}
+
+/// `IterParallel` is a parallel iterator that wraps a sequential iterator.
+///
+/// This type is created when using the `par_bridge` method on `ParallelBridge`. See the
+/// [`ParallelBridge`] documentation for details.
+///
+/// [`ParallelBridge`]: trait.ParallelBridge.html
+#[derive(Debug)]
+pub struct IterParallel<Iter> {
```

Comment:
Should we "bridge" this name too?  Maybe `IterBridge`, or even just `Bridge`?

Reply:
That would make more sense - this is still the same name it had from when the trait was named `AsParallel`. Between those two, I'm more of a fan of `IterBridge`.